iroh-bytes/src/store/bao_file.rs

```diff
@@ -0,0 +1,1238 @@
+        }
+        Ok(())
+    }
+
+    fn read_data_at(&self, offset: u64, len: usize) -> io::Result<Bytes> {
+        read_to_end(&self.data, offset, len)
+    }
+
+    fn read_outboard_at(&self, offset: u64, len: usize) -> io::Result<Bytes> {
+        read_to_end(&self.outboard, offset, len)
+    }
+}
+
+/// The storage for a bao file. This can be either in memory or on disk.
+#[derive(Debug)]
+pub(crate) enum BaoFileStorage {
+    /// The entry is incomplete and in memory.
+    ///
+    /// Since it is incomplete, it must be writeable.
+    ///
+    /// This is used mostly for tiny entries, <= 16 KiB. But in principle it
+    /// can be used for larger sizes.
+    ///
+    /// Incomplete mem entries are *not* persisted at all. So if the store
+    /// crashes they will be gone.
+    IncompleteMem(MutableMemStorage),
+    /// The entry is incomplete and on disk.
+    IncompleteFile(FileStorage),
+    /// The entry is complete. Outboard and data can come from different sources
+    /// (memory or file).
+    ///
+    /// Writing to this is a no-op, since it is already complete.
+    Complete(CompleteMemOrFileStorage),
+}
+
+impl Default for BaoFileStorage {
+    fn default() -> Self {
+        BaoFileStorage::Complete(Default::default())
+    }
+}
+
+impl BaoFileStorage {
+    /// Take the storage out, leaving an empty storage in its place.
+    ///
+    /// Be careful to put somethign back in its place, or you will lose data.
+    #[cfg(feature = "file-db")]
+    pub fn take(&mut self) -> Self {
+        std::mem::take(self)
+    }
+
+    /// Create a new mutable mem storage.
+    pub fn incomplete_mem() -> Self {
+        Self::IncompleteMem(Default::default())
+    }
+
+    /// Call sync_all on all the files.
+    fn sync_all(&self) -> io::Result<()> {
+        match self {
+            Self::Complete(_) => Ok(()),
+            Self::IncompleteMem(_) => Ok(()),
+            Self::IncompleteFile(file) => {
+                file.data.sync_all()?;
+                file.outboard.sync_all()?;
+                file.sizes.sync_all()?;
+                Ok(())
+            }
+        }
+    }
+
+    pub fn is_mem(&self) -> bool {
+        match self {
+            Self::IncompleteMem(_) => true,
+            Self::IncompleteFile(_) => false,
+            Self::Complete(c) => c.data.is_mem() && c.outboard.is_mem(),
+        }
+    }
+}
+
+#[cfg(test)]
+fn new_handle_id() -> u64 {
+    static HANDLE_ID: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);
+    HANDLE_ID.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+}
+
+/// A cheaply cloneable handle to a bao file, including the hash and the configuration.
+#[derive(Debug, Clone)]
+pub struct BaoFileHandle {
+    pub(crate) storage: Arc<RwLock<BaoFileStorage>>,
+    config: Arc<BaoFileConfig>,
+    hash: Hash,
+    #[cfg(test)]
+    pub(crate) id: u64,
+}
+
+impl Drop for BaoFileHandle {
+    fn drop(&mut self) {
+        if let Some(cb) = self.config.on_drop.as_ref() {
```

Comment:
Maybe let's change this so that the `OnDrop` callback is only called for the last drop? Don't think anyone using this would be interested in the drops before the last drop. Then the check in `file.rs:1458` can be removed, because it will only be called for the last_drop. Then you could use `Arc::get_mut` here to check if the last handle is being dropped, and only then call the `cb`.

Reply:
The issue is that we are interested in the second to last drop, because one entry is in the entry map in the actor. And just firing the cb on the second to last drop seems arbitrary.